owid/datautils/geo.py

```diff
@@ -1,3 +1,491 @@
 """Utils related to geographical entities.
 
 """
+
+import json
+import warnings
+from typing import List, Union, Set, Optional, Dict, Any
+
+import numpy as np
+import pandas as pd
+import owid.catalog as catalog
+
+from owid.datautils.common import ExceptionFromDocstring
+from owid.datautils.dataframes import groupby_agg
+
+# When creating region aggregates for a certain variable in a certain year, some mandatory countries must be
+# informed, otherwise the aggregate will be nan (since we consider that there is not enough information).
+# A country will be considered mandatory if they exceed this minimum fraction of the total population of the region.
+MIN_FRAC_INDIVIDUAL_POPULATION = 0.0
+# A country will be considered mandatory if the sum of the population of all countries (sorted by decreasing
+# population until reaching this country) exceeds the following fraction of the total population of the region.
+MIN_FRAC_CUMULATIVE_POPULATION = 0.7
+# Reference year to build the list of mandatory countries.
+REFERENCE_YEAR = 2018
+# Maximum fraction of nans allowed per year when doing aggregations (None to allow any fraction of nans).
+FRAC_ALLOWED_NANS_PER_YEAR = 0.2
+# Maximum number of nans allowed per year when doing aggregations (None to allow any number of nans).
+NUM_ALLOWED_NANS_PER_YEAR = None
+
+
+def _load_population() -> pd.DataFrame:
+    population = (
+        catalog.find("population", namespace="owid", dataset="key_indicators")
+        .load()
+        .reset_index()
+    )  # type: pd.DataFrame
+
+    return population
+
+
+def _load_countries_regions() -> pd.DataFrame:
+    countries_regions = catalog.find(
+        "countries_regions", dataset="reference", namespace="owid"
+    ).load()  # type: pd.DataFrame
+
+    return countries_regions
+
+
+def _load_income_groups() -> pd.DataFrame:
+    income_groups_found = catalog.find(
+        table="wb_income_group", dataset="wb_income", namespace="wb"
+    )
+    income_groups = (
+        income_groups_found[income_groups_found.path.str.startswith("garden")]
+        .load()
+        .reset_index()
+    )  # type: pd.DataFrame
+
+    return income_groups
+
+
+def _warn_on_list_of_entities(
+    list_of_entities: Union[List[str], Set[str]], warning_message: str, show_list: bool
```

Comment:
You could use `Sequence[str]` here as well. Up to you.

Reply:
Are you sure? When I use `list_of_entities: Sequence[str]`, then `list_of_entities` cannot be a `set` (I get an error about [line 480](https://github.com/owid/data-utils-py/blob/2ac06575a302a57b31cd8e88881bc89b312d7847/owid/datautils/geo.py#L480): `Argument "list_of_entities" to "_warn_on_list_of_entities" has incompatible type "Set[Any]"; expected "Sequence[str]"`).

Reply:
Sorry, you're right. I've mistakenly read it as `List` and `Tuple` 🤦